File: 74-search-a-2d-matrix/74-search-a-2d-matrix.c
bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target){
    int lastElement = *matrixColSize-1;
    int l,r,mid;
    
    for(int i = 0; i < matrixSize; i++){
        if(target <= matrix[i][lastElement]){   //if (target > matrix[i][lastElement]), just i++;
            l = 0;
            r = *matrixColSize-1;
            while(l <= r){      //binary search
                mid = (l+r)/2;
                if(target > matrix[i][mid])
                    l = mid+1;
                else if(target < matrix[i][mid])
                    r = mid-1;
                else
                    return true;
            }
        }
    }
    return false;
}

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int len = matrix[0].length;
        int l = 0, r = len-1, mid;
        int i = 0;
        
        while(l <= r && i < matrix.length){
            if(matrix[i][r] < target)
                i++;
            else{
                mid = (l+r)/2;
                if(matrix[i][mid] == target)
                    return true;
                else if(matrix[i][mid] < target)
                    l++;
                else
                    r--;
            }
        }
        return false;
    }
}